forest/cli/src/cli/send_cmd.rs

```diff
@@ -7,9 +7,88 @@ use forest_json::message::json::MessageJson;
 use forest_rpc_client::{mpool_push_message, wallet_default_address};
 use fvm_shared::{address::Address, econ::TokenAmount, message::Message, METHOD_SEND};
 use num::BigInt;
+use rust_decimal::prelude::*;
+use rust_decimal_macros::dec;
 
 use super::{handle_rpc_err, Config};
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FILAmount {
+    pub value: TokenAmount,
+}
+
+impl FromStr for FILAmount {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let suffix_idx = s.rfind(char::is_numeric);
+        let (val, suffix) = match suffix_idx {
+            Some(idx) => s.split_at(idx + 1),
+            None => return Err(anyhow::anyhow!("failed to parse string: {}", s)),
+        };
+        let suffix = suffix.replace(' ', "");
+
+        let mut multiplier = dec!(1.0);
+        let prefix = if !suffix.is_empty() {
+            match suffix.trim().to_lowercase().strip_suffix("fil") {
+                Some("atto" | "a") => "atto",
+                Some("femto") => {
+                    multiplier *= dec!(1_000);
+                    "femto"
+                }
+                Some("pico") => {
+                    multiplier *= dec!(1_000_000);
+                    "pico"
+                }
+                Some("nano") => {
+                    multiplier *= dec!(1_000_000_000);
+                    "nano"
+                }
+                Some("micro") => {
+                    multiplier *= dec!(1_000_000_000_000);
+                    "micro"
+                }
+                Some("milli") => {
+                    multiplier *= dec!(1_000_000_000_000_000);
+                    "milli"
+                }
+                Some("" | " ") => {
+                    multiplier *= dec!(1_000_000_000_000_000_000);
+                    ""
+                }
+                _ => {
+                    return Err(anyhow::anyhow!("unrecognized suffix: {}", suffix));
+                }
+            }
+        } else {
+            multiplier *= dec!(1_000_000_000_000_000_000);
+            ""
+        };
+
+        if val.chars().count() > 50 {
+            anyhow::bail!("string length too large: {}", val.chars().count());
+        }
+
+        let parsed_val = Decimal::from_str(val)
+            .map_err(|_| anyhow::anyhow!("failed to parse {} as a decimal number", val))?;
+
+        let attofil_val = if (parsed_val * multiplier).fract() != dec!(0.0) {
+            return Err(anyhow::anyhow!("invalid {}FIL value: {}", prefix, val));
```

Comment:
Here you can just say that the FIL amount has to be a whole number of atto FIL. This lets you remove the `prefix` variable which will simplify your code a lot.